File: killrvideo-service-users/src/main/java/com/killrvideo/service/user/grpc/UserManagementServiceGrpcMapper.java
package com.killrvideo.service.user.grpc;

import com.google.protobuf.Timestamp;
import com.killrvideo.service.user.dto.User;
import killrvideo.user_management.UserManagementServiceOuterClass.CreateUserRequest;
import killrvideo.user_management.UserManagementServiceOuterClass.UserProfile;
import killrvideo.user_management.UserManagementServiceOuterClass.VerifyCredentialsResponse;
import killrvideo.user_management.events.UserManagementEvents;
import org.springframework.stereotype.Component;

import java.time.Instant;
import java.util.UUID;

import static com.killrvideo.utils.GrpcMappingUtils.uuidToUuid;

/**
 * Mapping from interfaces GRPC to DTO
 *
 * @author Cedrick LUNVEN (@clunven)
 */
@Component
public class UserManagementServiceGrpcMapper {

    public User mapUserRequest2User(CreateUserRequest grpcReq) {
        User user = new User();
        user.setEmail(grpcReq.getEmail());
        user.setCreatedDate(Instant.now());
        user.setFirstname(grpcReq.getFirstName());
        user.setLastname(grpcReq.getLastName());
        user.setUserid(UUID.fromString(grpcReq.getUserId().getValue()));
        return user;
    }
    
    public UserProfile mapUserToGrpcUserProfile(User user) {
       return UserProfile.newBuilder()
                    .setEmail(user.getEmail())
                    .setFirstName(user.getFirstname())
                    .setLastName(user.getLastname())
                    .setUserId(uuidToUuid(user.getUserid()))
                    .build();
    }
    
    public VerifyCredentialsResponse mapResponseVerifyCredentials(UUID userid) {
        return VerifyCredentialsResponse.newBuilder().setUserId(uuidToUuid(userid)).build();
    }

    public UserManagementEvents.UserCreated createUserCreatedEvent(User user) {
        return UserManagementEvents.UserCreated.newBuilder()
                .setEmail(user.getEmail())
                .setFirstName(user.getFirstname())
                .setLastName(user.getLastname())
                .setUserId(uuidToUuid(user.getUserid()))
                .setTimestamp(Timestamp.newBuilder().build())
                .build();
    }
}

File: killrvideo-service-users/src/main/kotlin/com/killrvideo/service/user/grpc/UserManagementServiceGrpcMapper.kt
package com.killrvideo.service.user.grpc

import com.google.protobuf.Timestamp
import com.killrvideo.service.user.dto.User
import com.killrvideo.utils.GrpcMappingUtils.fromUuid
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import com.killrvideo.utils.HashUtils
import killrvideo.user_management.UserManagementServiceOuterClass.CreateUserRequest
import killrvideo.user_management.UserManagementServiceOuterClass.GetUserProfileRequest
import killrvideo.user_management.events.userCreated
import killrvideo.user_management.getUserProfileResponse
import killrvideo.user_management.userProfile
import killrvideo.user_management.verifyCredentialsResponse
import org.apache.commons.lang3.StringUtils.isNotBlank
import org.springframework.stereotype.Component
import java.time.Instant
import java.util.*

/**
 * Mapping from interfaces GRPC to DTO
 *
 * @author Cedrick LUNVEN (@clunven)
 */
@Component
open class UserManagementServiceGrpcMapper {
    object CreateUserRequestExtensions {
        fun CreateUserRequest.parse(): Pair<User, String> =
            Pair(
                User(
                    userid = fromUuid(this.userId),
                    email = this.email,
                    firstname = this.firstName,
                    lastname = this.lastName,
                    createdDate = Instant.now()
                ),
                HashUtils.hashPassword(this.password.trim())
            )
    }

    object GetUserProfileRequestExtensions {
        fun GetUserProfileRequest.parse(): List<UUID> =
            this.userIdsList
                .filter { isNotBlank(it.value) }
                .map { fromUuid(it) }
    }

    fun mapResponseVerifyCredentials(userid: UUID?) =
        verifyCredentialsResponse {
            userid?.let { userId = uuidToUuid(it) }
        }

    fun createUserCreatedEvent(user: User) =
        userCreated {
            user.email?.let {email = it}
            user.firstname?.let {firstName = it}
            user.lastname?.let {lastName = it}
            user.userid?.let {userId = uuidToUuid(it)}
            timestamp = Timestamp.newBuilder().build()
        }

    fun buildGetUserProfileResponse(users: List<User>) =
        getUserProfileResponse {
            profiles.addAll(users.map { mapUserToGrpcUserProfile(it) })
        }

    private fun mapUserToGrpcUserProfile(user: User) =
        userProfile {
            user.email?.let { email = it }
            user.firstname?.let { firstName = it }
            user.lastname?.let { lastName = it }
            user.userid?.let { userId = uuidToUuid(it) }
        }
}
